File: bms/go/triangolatte/isEar/isEar.go
package test

var ()

const ()

func isEar(p *Element) bool {
	a, b, c := p.Prev.Point, p.Point, p.Next.Point
	if isReflex(a, b, c) {
		return false
	}

	r := p.Next.Next
	for r != p.Prev {
		inside := isInsideTriangle(a, b, c, r.Point)
		reflex := isReflex(r.Prev.Point, r.Point, r.Next.Point)
		if inside && reflex {
			return false
		}
		r = r.Next
	}
	return true
}

func isReflex(a, b, c Point) bool {
	return (b.X-a.X)*(c.Y-b.Y)-(c.X-b.X)*(b.Y-a.Y) < 0
}

func isInsideTriangle(a, b, c, p Point) bool {
	return (c.X-p.X)*(a.Y-p.Y)-(a.X-p.X)*(c.Y-p.Y) >= 0 &&
		(a.X-p.X)*(b.Y-p.Y)-(b.X-p.X)*(a.Y-p.Y) >= 0 &&
		(b.X-p.X)*(c.Y-p.Y)-(c.X-p.X)*(b.Y-p.Y) >= 0
}

type Element struct {
	Prev, Next *Element
	Point      Point
}

type Point struct {
	X, Y float64
}

File: translations_only/fail/go/triangolatte/isEar/mistral/Hinted/isEar.rs


use std::mem;
use std::option::Option::None;
use std::option::Option::Some;

#[derive(PartialEq)]
struct Element {
    prev: Option<Box<Element>>,
    next: Option<Box<Element>>,
    point: Point,
}

#[derive(PartialEq)]
struct Point {
    x: f64,
    y: f64,
}

fn is_ear(p: &Element) -> bool {
    let a = &p.prev.as_ref().unwrap().point;
    let b = &p.point;
    let c = &p.next.as_ref().unwrap().point;
    if is_reflex(a, b, c) {
        return false;
    }

    let mut r = &*p.next.as_ref().unwrap().next.as_ref().unwrap();
    while r != &*p.prev.as_ref().unwrap() {
        let inside = is_inside_triangle(a, b, c, &r.point);
        let reflex = is_reflex(&r.prev.as_ref().unwrap().point, &r.point, &r.next.as_ref().unwrap().point);
        if inside && reflex {
            return false;
        }
        r = &*r.next.as_ref().unwrap();
    }
    true
}

fn is_reflex(a: &Point, b: &Point, c: &Point) -> bool {
    (b.x - a.x) * (c.y - b.y) - (c.x - b.x) * (b.y - a.y) < 0.0
}

fn is_inside_triangle(a: &Point, b: &Point, c: &Point, p: &Point) -> bool {
    let check_side = |a: &Point, b: &Point, p: &Point| (b.x - p.x) * (a.y - p.y) - (a.x - p.x) * (b.y - p.y);
    check_side(c, a, p) >= 0.0 && check_side(a, b, p) >= 0.0 && check_side(b, c, p) >= 0.0
}

fn main() {
    // Test cases go here
}

